metomi/rose/loc_handlers/git.py

```diff
@@ -0,0 +1,168 @@
+            and not os.path.exists(loc.name)  # same as svn...
+            and not self.manager.popen.run(
+                "git", "ls-remote", "--exit-code", remote)[0]
+            # https://superuser.com/questions/227509/git-ping-check-if-remote-repository-exists
+        )
+
+    def parse(self, loc, conf_tree):
+        """Set loc.real_name, loc.scheme, loc.loc_type.
+
+        Within Git we have a lot of trouble figuring out remote
+        loc_type - a clone is required, unfortunately. There is a
+        tradeoff between extra Git commands and bandwidth vs
+        parse failure behaviour. We have decided to short cut the
+        loc_type calculation to save commands and bandwidth,
+        catching failures later.
+
+        """
+        loc.scheme = self.SCHEMES[0]
+        remote, path, ref = self._parse_name(loc)
+
+        # Extract the commit hash if we don't already have it.
+        commithash = self._get_commithash(remote, ref)
+
+        if path.endswith("/"):  # This is a short cut, checked later.
+            loc.loc_type = loc.TYPE_TREE
+        else:
+            loc.loc_type = loc.TYPE_BLOB
+        loc.real_name = (
+            f"remote:{remote} ref:{ref} commit:{commithash} path:{path}"
+        )
+        loc.key = commithash
+
+    async def pull(self, loc, conf_tree):
+        """Get loc to its cache.
+
+        git sparse-checkout is not available below Git 2.25, and seems to omit
+        contents altogether if set to the root of the repo (as of 2.40.1).
+
+        Filtering requires uploadpack.allowFilter to be set true on
+        the remote repo or server.
+
+        """
+        if not loc.real_name:
+            self.parse(loc, conf_tree)
+        remote, path, ref = self._parse_name(loc)
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            git_dir_opt = f"--git-dir={tmpdirname}/.git"
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, "init"
+            )
+            if self.git_version >= (2, 25, 0) and path != "./":
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "sparse-checkout", "set", path,
+                    "--no-cone"
+                )
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1",
+                    "--filter=blob:none", remote, loc.key
+                )
+            else:
+                await self.manager.popen.run_ok_async(
+                    "git", git_dir_opt, "fetch", "--depth=1", remote, loc.key
+                )
+
+            await self.manager.popen.run_ok_async(
+                "git", git_dir_opt, f"--work-tree={tmpdirname}", "checkout",
+                loc.key
+            )
+            name = tmpdirname + "/" + path
+
+            # Check that we have inferred the right type from the path name.
+            real_loc_type = loc.TYPE_TREE if os.path.isdir(name) else loc.TYPE_BLOB
```

Comment:
This should make the linter happy:

```suggestion
            real_loc_type = (
                loc.TYPE_TREE if os.path.isdir(name) else loc.TYPE_BLOB
            )
```